### PR TITLE
ipq40xx: Add support for Linksys EA6350v3

### PR DESCRIPTION
**Pull request by Escalion**
Specifications:
SOC:    Qualcomm IPQ4018
RAM:    256 MiB Samsung K4B2G1646F-BYK0
FLASH1: MX25L1605D 2 MB
FLASH2:  Winbond W25N01GV 128Mb
ETH:    Qualcomm QCA8075
WLAN1:  Qualcomm Atheros QCA4018 2.4GHz 802.11b/g/n 2x2
WLAN2:  Qualcomm Atheros QCA4018 5GHz 802.11n/ac W2 2x2
INPUT:    WPS, Reset
LED:           Status - Green
SERIAL: Header at J19, Beneath DC Power Jack
                1-VCC ; 2-TX ; 3-RX; 4-GND;
                Serial 115200-8-N-1.
Tested and working:
 - USB
 - LAN Ethernet (Correct MAC-address)
 - WAN Ethernet (Correct MAC-address)
 - 2.4 GHz WiFi (Correct MAC-address)
 - 5 GHz WiFi (Correct MAC-address)
 - Factory installation from Web UI
 - OpenWRT sysupgrade
 - LED
 - WPS Button (No event attached - logread method)
 - Reset Button

Not Working:
 - No known issues

Install via Web UI:

 - Attach to a LAN port on the router.
 - Connect to the Linksys Smart WiFi Page (default 192.168.1.1)
 - Select the connectivity tab on the left
 - In the manual update box on the right;
 - Select browse, and browse to openwrt-ipq40xx-linksys_ea6350v3-squashfs-factory.bin
 - Click update.
 - Read and accept the warning (or not, if you don't want to take the risk)
 - The router LED will start blinking. When the router LED goes solid, you can now navigate to 192.168.1.1 to your new OpenWRT installation.

Recovery (Automatic):
 - If the device fails to boot after install or upgrade, whilst the unit is turned on:
 1 - Wait 1 minute
 2 - Switch Off and Wait 10 seconds
 3 - Switch on
 4 - Repeat steps 1 to 3, 3 times then go to 5.
 5 - U-boot will have now erased the failed update and switched back to the last working firmware - you should be able to access your router on LAN at 192.168.1.1

Recovery (Manual):
- The steps for manual recovery are the same as the generic u-boot tftp client method.

Signed-off-by: Ryan Pannell <ryan@osukl.com>

### CODE REVIEW

File: target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ea6350v3.dts [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+			gpios = <&tlmm 0 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		power: status {
+			label = "rt-ac58u:blue:status";
```
Comment (robimarko) [inline target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ea6350v3.dts]: Wrong name, probably left over

**Comment by pkgadd** [comment]
Please consider ammending your commit message with something like https://github.com/openwrt/openwrt/commit/4b280ad91a6b155ac71c417aaa7bb5f4e328712f, to get the basic device info and flash instructions into the git log.

File: target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ea6350v3.dts (hunk 1 above)
Comment (Escalion) [reply target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ea6350v3.dts]: Resolved

**Comment by Escalion** [comment]
> Please consider ammending your commit message with something like [4b280ad](https://github.com/openwrt/openwrt/commit/4b280ad91a6b155ac71c417aaa7bb5f4e328712f), to get the basic device info and flash instructions into the git log.

Amended.

**Comment by Escalion** [comment]
@diizzyy there is - on the third failed boot the bootloader will revert back to the previously used image or you can use the serial + tftp method. Should this be documented in the commit?

**Comment by diizzyy** [comment]
@Escalion 
Yes, please

**Comment by chunkeey** [comment]
@Escalion 
can you please squash all the patches in this PR into a single one? thanks.

**Comment by diizzyy** [comment]
Awesome, so it has some kind of dual firmware "partition" failover as far as I can tell by the above?

**Comment by Escalion** [comment]
@diizzy that is correct. It has a kernel / rootfs and alt_kernel / alt_rootfs partition scheme. It uses s_env to store a signature which is in the format of int32 magic, int32 bootcount, int32 checksum.
U-boot increments the bootcount and If the bootcount is not reset to 0, when the bootcount hits 3 it will wipe the failed partition and revert to booting the other partition.

File: package/boot/uboot-envtools/files/ipq40xx [hunk 2]
```diff
@@ -35,6 +35,9 @@ openmesh,a42 |\
 openmesh,a62)
 	ubootenv_add_uci_config "/dev/mtd5" "0x0" "0x10000" "0x10000"
 	;;
+linksys,ea6350v3)
+	ubootenv_add_uci_config "/dev/mtd7" "0x0" "0x20000" "0x20000"
+	;;
```
Comment (blogic) [inline package/boot/uboot-envtools/files/ipq40xx]: please order these alphabetically

File: include/image-commands.mk [hunk 3]
```diff
@@ -344,3 +344,10 @@ define Build/kernel2minor
 	kernel2minor -k $@ -r $@.new $(1)
 	mv $@.new $@
 endef
+
+define Build/linksys-image
+	sh $(TOPDIR)/scripts/linksys-image.sh \
+		"$(call param_get_default,type,$(1),$(DEVICE_NAME))" \
```
Comment (blogic) [inline include/image-commands.mk]: why is there and explicit sh call invoking a shell script ?

File: package/system/mtd/src/Makefile [hunk 4]
```diff
@@ -15,6 +15,7 @@ obj.ramips = $(obj.seama) $(obj.wrg)
 obj.mvebu = linksys_bootcount.o
 obj.kirkwood = linksys_bootcount.o
 obj.ipq806x = linksys_bootcount.o
+obj.ipq40xx = linksys_bootcount.o
```
Comment (blogic) [inline package/system/mtd/src/Makefile]: this needs to go into its own patch

File: package/firmware/ipq-wifi/Makefile [hunk 5]
```diff
@@ -17,7 +17,7 @@ endef
 # Please send a mail with your device-specific board files upstream.
```
Comment (blogic) [inline package/firmware/ipq-wifi/Makefile]: this needs to go into its own patch

File: target/linux/ipq40xx/base-files/etc/board.d/02_network [hunk 6]
```diff
@@ -25,6 +25,14 @@ asus,rt-ac58u)
 	ucidef_set_interface_macaddr "lan" "$lan_mac_addr"
 	ucidef_set_interface_macaddr "wan" "$wan_mac_addr"
 	;;
+linksys,ea6350v3 )
+	hw_mac_addr=$(mtd_get_mac_ascii devinfo hw_mac_addr)
```
Comment (blogic) [inline target/linux/ipq40xx/base-files/etc/board.d/02_network]: this is not how the alphabet works

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 7]
```diff
@@ -120,6 +120,7 @@ case "$FIRMWARE" in
 "ath10k/pre-cal-ahb-a000000.wifi.bin")
 	case "$board" in
 	8dev,jalapeno |\
+	linksys,ea6350v3 |\
 	glinet,gl-b1300 |\
```
Comment (blogic) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: ...

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 8]
```diff
@@ -156,6 +157,7 @@ case "$FIRMWARE" in
 "ath10k/pre-cal-ahb-a800000.wifi.bin")
 	case "$board" in
 	8dev,jalapeno |\
+	linksys,ea6350v3 |\
```
Comment (blogic) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: .....

File: target/linux/ipq40xx/base-files/lib/upgrade/platform.sh [hunk 9]
```diff
@@ -49,6 +49,9 @@ platform_do_upgrade() {
 	8dev,jalapeno)
 		nand_do_upgrade "$ARGV"
 		;;
+	linksys,ea6350v3)
+		platform_do_upgrade_linksys "$ARGV"
```
Comment (blogic) [inline target/linux/ipq40xx/base-files/lib/upgrade/platform.sh]: ...

File: target/linux/ipq40xx/image/Makefile [hunk 10]
```diff
@@ -45,7 +45,6 @@ define Device/DniImage
 endef
 DEVICE_VARS += NETGEAR_BOARD_ID NETGEAR_HW_ID
 
-
 define Device/8dev_jalapeno
```
Comment (blogic) [inline target/linux/ipq40xx/image/Makefile]: unrelated change

File: target/linux/ipq40xx/patches-4.14/997-device_tree_cmdline.patch [hunk 11]
```diff
@@ -1,12 +1,15 @@
...
-@@ -1130,6 +1130,9 @@ int __init early_init_dt_scan_chosen(uns
+@@ -1130,6 +1130,12 @@ int __init early_init_dt_scan_chosen(uns
  	p = of_get_flat_dt_prop(node, "bootargs", &l);
  	if (p != NULL && l > 0)
  		strlcpy(data, p, min((int)l, COMMAND_LINE_SIZE));
 +	p = of_get_flat_dt_prop(node, "bootargs-append", &l);
 +	if (p != NULL && l > 0)
 +		strlcat(data, p, min_t(int, strlen(data) + (int)l, COMMAND_LINE_SIZE));
- 
++	p = of_get_flat_dt_prop(node, "bootargs-override", &l);
```
Comment (blogic) [inline target/linux/ipq40xx/patches-4.14/997-device_tree_cmdline.patch]: needs to go into its own patch, explaining why its needed

File: include/image-commands.mk (hunk 3 above)
Comment (Escalion) [reply include/image-commands.mk]: It doesn't need to be explicit. It could just be called as $(TOPDIR)/scripts/linksys-image.sh.
Will remove 'sh'

**Comment by mkresin** [comment]
Please reopen the PR if you have the time to address the requested changes.
